OCD-3263: Adjust listing validation for ERD-Phase-2 and Grace Period

kekey1 opened the pull request.

kekey1 committed:
fix: Compare newly added criteria by id rather than number

Since the ERD-Phase-2 removed criteria all have an identically-numbered
counterpart, the RemovedCriteriaComparison reviewer was incorrectly
comparing updated vs existing criteria. Switching the criteria check to
use the id should resolve that confusion.

[#OCD-3263]

kekey1 committed:
feat!: Revise required/related rules after grace period

The required and related validation needs to not consider removed
criteria. Additionally, after a "grace period" identified by a flag,
removed criteria can no longer be considered to meet the requirement.
For example, pre-ERD-grace period, criterion F1 required either D2 or
D2Cures. Post-ERD-grace-period, F1 must have D2Cures.

An entirely separate reviewer is written to run based on flag
state. The new reviewer makes less use of the ValidationUtils
class which has some methods that check criteria by number and
is working oddly now that some removed criteria are those that
have similarly-numbered Cures criteria.

[#OCD-3263]

kekey1 committed:
feat!: Only check active criteria for d12/d13 attestation requirement

Some of the criteria that currently trigger the d12/d13
requirement are getting removed as of ERD-Phase-2. We should no longer
consider those removed criteria when calculating Privacy and Security
validation.

[#OCD-3263]

kekey1 committed:
feat: Consider only active criteria when checking for CQM reqs

c3-original is being removed. If c3-original is attested, it should no
longer require that at least one CQM references it.

[#OCD-3263]

kekey1 committed:
feat!: Revise required/related rules during grace period

During the grace period, there will be no rules enforcing removed
criteria to have any required/related criteria. However, removed
criteria (i.e. the "original" criteria with newer Cures Update versions)
will continue to count towards meeting required/related criteria
validation rules. This allows users to continue editing listings for a
period of time.

The grace period is determined by a flag.

[#OCD-3263]

kekey1 committed:
feat!: Stop giving listing validation warnings for removed criteria

[#OCD-3841]

kekey1 committed:
Merge remote-tracking branch 'upstream/staging' into OCD-3263

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
Merge remote-tracking branch 'upstream/staging' into OCD-3263

kekey1 committed:
feat: Adjust P&S validation during and after grace period

Before Dec 31 and during the grace period, the P&S validation should
work as it currently does, aside from not considering removed criteria.

After the grace period, P&S Validation for listing edit should work
similarly to the way it does on listing upload. It should no longer
care if a criteria is being added.

[#OCD-3263]

kekey1 committed:
feat: Remove the P&S validation check around certification date

Cures Rule-Effective-Date was 2 years ago and we no longer need to check
if a listing that is being uploaded was certified before or after that.

[#OCD-3263]

andlar commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/Edition2015ListingValidator.java:
```diff
@@ -224,64 +241,73 @@ public class Edition2015ListingValidator extends Validator {
+        comparisonReviewers.add(ucdCriteriaComparisonReviewer);
+        comparisonReviewers.add(testFunctionalityAllowedByRoleReviewer);
+        comparisonReviewers.add(listingStatusAndUserRoleReviewer);
+        if (!ff4j.check(FeatureList.ERD_PHASE_2)
+                || (ff4j.check(FeatureList.ERD_PHASE_2) && ff4j.check(FeatureList.ERD_PHASE_2_GRACE_PERIOD))) {
+            comparisonReviewers.add(privacyAndSecurityCriteriaReviewerPreErdPhase2);
```
This looks like we're adding the "p&sCriteria Pre-ERD-Phase-2" in the time before ERD-Phase-2 and after ERD-Phase-2-Grace-Period, but not during the grace period. Is that right? So it would add it today, and in June of 2023, but not in Feb of 2023?

kekey1 replied:
Hmmm. I mean for it to apply 1) now, before Dec 31 - so ERD-Phase-2 is off, and 2) During the grace period - so ERD-Phase-2 is ON and the grace-period flag is ON.  That is how I thought the flags would work. On Dec-31 we will turn on BOTH flags. I don't think it would be a valid state to have the grace period flag on with the ERD-Phase-2 flag off. So this validator should run now and feb 2023 but not june 2023. In June 2023, the other p&s reviewer should run but it should never run before that (higher up in the diff). Can you double-check my logic?

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
test: Remove non-functional test

[#OCD-3263]

kekey1 committed:
feat: Change usage of grace-period flag to grace-period-end

[#OCD-3263-correct-grace-periods]

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
fix: Do not check old criteria after erd-phase-2. They are removed.

An error is already given on listing upload that a listing cannot attest
to removed criteria. We do not need an additional error from the
OldCriteriaWithoutIcsReviewer.

[#OCD-3263]

kekey1 committed:
fix: The grace period applies to listing edit. Listing upload has none.

The grace period, allowing of old/removed criteria to meet criteria
dependency requirement, is only for existing listings. This update
removes the notion of a grace period from upload.

[#OCD-3263]

kekey1 committed:
fix: Do not show conformance method warnings for removed criteria

[#OCD-3841]

kekey1 committed:
fix: Remove invalid CQM Criteria and handle c3 better

If a listing upload uses "c3" as a CQM Criteria but doesn't attest to
c3-original or c3-cures, we will pick c3-cures for them so they can make
appropriate edits in the UI

[#OCD-3263]

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
fix: Allow either c3 criteria to suffice for CQMs

Future ticket to enforce that the CQM c3 matches the listing c3.

[#OCD-3263]

kekey1 committed:
Merge branch 'staging' into OCD-3263

kekey1 committed:
fix: Handle null conformance methods in reviewer

This occurs if the CM version is provided in the upload file but a CM
name is not provided. If there is more than 1 CM available for the
criteria, then a suitable default cannot be determined and the CM
remains null.

[#OCD-3263]

kekey1 committed:
fix: Add email sending to upload job for processing errors

An email has always been sent from the ListingUploadController, and
still gets sent from there in the event that any required headings are
missing and we cannot proceed with the file. However, after moving the
initial upload validation to a quartz job we lost the feature of sending
an email when there was any sort of error parsing, normalizing or
validating the file.

[#OCD-3263]

kekey1 committed:
fix: Eliminate errors on removed criteria test tools during listing edit

[#OCD-3263]

kekey1 committed:
Merge branch 'staging' into OCD-3263
